k8s.gcr.io/lib.sh

```diff
@@ -0,0 +1,279 @@
+        return 1
+    fi
+    project="$1"
+    group="$2"
+    region="${3:-}"
+
+    gsutil iam ch \
+        "group:${group}:objectAdmin" \
+        $(gcs_bucket_for "${project}" "${region}")
+    gsutil iam ch \
+        "group:${group}:legacyBucketReader" \
+        $(gcs_bucket_for "${project}" "${region}")
+}
+
+# Grant full privileges to the GCR promoter bot
+# $1: The GCP project
+function empower_promoter() {
+    if [ $# -lt 1 -o $# -gt 2 -o -z "$1" ]; then
+        echo "empower_promoter(project, [region]) requires 1 or 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    region="${2:-}"
+
+    acct=$(svc_acct_for "${project}" "${PROMOTER_SVCACCT}")
+
+    if ! gcloud --project "${project}" iam service-accounts describe "${acct}" >/dev/null 2>&1; then
+        gcloud --project "${project}" \
+            iam service-accounts create \
+            "${name}" \
+            --display-name="k8s-infra container image promoter"
+    fi
+
+    # Grant admins access to do admin stuff.
+    gsutil iam ch \
+        "serviceAccount:${acct}:objectAdmin" \
+        $(gcs_bucket_for "${project}" "${region}")
+    gsutil iam ch \
+        "serviceAccount:${acct}:legacyBucketOwner" \
+        $(gcs_bucket_for "${project}" "${region}")
+}
+
+# Configure bucket lifecycle for a prod repo
+# $1: The GCP project
+function ensure_prod_lifecycle() {
+    if [ $# -lt 1 -o $# -gt 2 -o -z "$1" ]; then
+        echo "ensure_prod_lifecycle(project, [region]) requires 1 or 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    region="${2:-}"
+
+    # Set lifecycle policies.
+    # TODO: do we want this?  It will inhibit promoter's GC, but will protect
+    # against mistakenly nuking images we wanted to keep.
+    #TODO: if we keep this, maybe we want to lock this (can't be overridden)?
+    gsutil retention set 5y $(gcs_bucket_for "${project}" "${region}")
+}
+
+# Configure bucket lifecycle for a staging repo
+# $1: The GCP project
+function ensure_staging_lifecycle() {
+    if [ $# -lt 1 -o $# -gt 2 -o -z "$1" ]; then
+        echo "ensure_staging_lifecycle(project, [region]) requires 1 or 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    region="${2:-}"
+
+    # Set lifecycle policies.
+    gsutil retention set 30d $(gcs_bucket_for "${project}" "${region}")
+    echo '
+        {
+          "rule": [
+            {
+              "condition": {
+                "age": 30
+              },
+              "action": {
+                "storageClass": "NEARLINE",
+                "type": "SetStorageClass"
+              }
+            },
+            {
+              "condition": {
+                "age": 90
+              },
+              "action": {
+                "type": "Delete"
+              }
+            }
+          ]
+        }
+        ' | gsutil lifecycle set /dev/stdin $(gcs_bucket_for "${project}" "${region}")
```

Comment:
Hmm neat trick with the echo + /dev/stdin there.

Reply:
falls victim to the same non-portability you called out above, but I couldn't get it to parse EOL properly with a HERE doc, and gave up.  I could write it to a file if it matters.

Reply:
Hmm true re: non-portability.

heredoc issues aside, is there a threat of /dev/stdin becoming corrupted due to human input while this script is running...? If so we should just write to a temporary file.

Reply:
no, because the | explicitly maps /dev/stdin to that contents via a pipe

/dev/stdin is supported on linux and mac, which is all I care about for this script.  I will revisit a heredoc later if needed